server/src/test/java/org/elasticsearch/cluster/routing/ShardRoutingTests.java

```diff
@@ -11,16 +11,127 @@
 import org.elasticsearch.Version;
 import org.elasticsearch.common.UUIDs;
 import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
 import org.elasticsearch.index.Index;
 import org.elasticsearch.index.shard.ShardId;
 import org.elasticsearch.repositories.IndexId;
 import org.elasticsearch.snapshots.Snapshot;
 import org.elasticsearch.snapshots.SnapshotId;
+import org.elasticsearch.test.AbstractWireSerializingTestCase;
 import org.elasticsearch.test.ESTestCase;
 
 import java.io.IOException;
 
-public class ShardRoutingTests extends ESTestCase {
+import static java.util.Objects.requireNonNullElseGet;
+import static org.hamcrest.Matchers.containsString;
+
+public class ShardRoutingTests extends AbstractWireSerializingTestCase<ShardRouting> {
+
+    @Override
+    protected Writeable.Reader<ShardRouting> instanceReader() {
+        return ShardRouting::new;
+    }
+
+    @Override
+    protected ShardRouting createTestInstance() {
+        ShardRoutingState state = randomFrom(ShardRoutingState.values());
+        return TestShardRouting.newShardRouting(
+            randomIdentifier(),
+            randomIntBetween(0, 9),
+            state == ShardRoutingState.UNASSIGNED ? null : randomIdentifier(),
+            state == ShardRoutingState.RELOCATING ? randomIdentifier() : null,
+            state != ShardRoutingState.UNASSIGNED && randomBoolean(),
+            state
+        );
+    }
+
+    @Override
+    protected ShardRouting mutateInstance(ShardRouting instance) throws IOException {
+        return switch (randomInt(3)) {
+            case 0 -> mutateShardId(instance);
+            case 1 -> mutateCurrentNodeId(instance);
+            case 2 -> mutateState(instance);
+            default -> randomValueOtherThan(instance, this::createTestInstance);
+        };
+    }
+
+    private static ShardRouting mutateShardId(ShardRouting instance) {
+        var newShardId = switch (randomInt(2)) {
+            case 0 -> new ShardId(
+                randomValueOtherThan(instance.getIndexName(), ESTestCase::randomIdentifier),
+                instance.shardId().getIndex().getUUID(),
+                instance.id()
+            );
+            case 1 -> new ShardId(
+                instance.getIndexName(),
+                randomValueOtherThan(instance.shardId().getIndex().getUUID(), UUIDs::randomBase64UUID),
+                instance.id()
+            );
+            case 2 -> new ShardId(
+                instance.getIndexName(),
+                instance.shardId().getIndex().getUUID(),
+                randomValueOtherThan(instance.id(), () -> randomIntBetween(0, 99))
+            );
+            default -> throw new RuntimeException("unreachable");
+        };
+        return TestShardRouting.newShardRouting(
```

Comment:
Hmm I think these shorthand `newShardRouting` methods that don't specify anything will generate too many mutations. For instance if the shard is unassigned then you'll get a new random `unassignedInfo` here.

Reply:
Updated mutations to reuse values when available and fallback to `TestShardRouting` for generating new ones otherwise